keps/draft-20180426-security-conformance.md

```diff
@@ -0,0 +1,532 @@
+
+  *Conformance Profile: Essential*
+
+  Every namespace has a service account named `default`.
+  Without explicit permission grant from RBAC RoleBinding or ClusterRoleBinding,
+  This service account has the minimum privilege to access Kubernetes API,
+  it's forbidden for the service account to `get/list/watch/create/update/patch/delete` any of API objects
+
+###### Test Recommendation
+
+  The test code consumes Kubernetes API as a serviceaccount/user with privileges:
+
+  - Create/Delete a namespace
+  - Impersonate the `default` service account in the new namespace
+
+
+  Test setup
+
+  - Create a new namespace
+  - Create a Pod without specifying `SecurityContext`, `ServiceAccountName`, without any volumes
+
+
+  Test steps  
+
+  - Impersonate the `default` service account to
+    `get/list/watch/create` the objects defined by `core` api group and get _forbidden_
+  - Impersonate the `default` service account to
+    `get/list/watch/create` the rbac objects: `Role/RoleBinding/ClusterRole/ClusterRoleBinding` and get _forbidden_
+  - Impersonate the `default` service account to
+    `update/patch` the created Pod by changing some fields, like a label, should get _forbidden_
+  - Impersonate the `default` service account to `delete` the Pod, and should get _forbidden_
+
+
+  Test teardown
+
+  - Delete the namespace
+
+#### Kubelet r/w API Access
+
+  *Conformance Profile: Essential*
+
+  _TODO: need more context_
+
+  Anonymous access to Kubelet r/w APIs should be denied as unauthorized.
+  Default service accounts should also be denied as forbidden accessing Kubelet r/w APIs.
+
+#### Privileged Containers
+
+  *Conformance Profile: Restricted*
+
+  Pod creation with `privileged` set to `true` in `SecurityContext` (Pod or container level)
+  should fail if there's no explicit RBAC associated with the Pod service account
+  (`default` or explicitly specified in `ServiceAccountName`).
+
+###### Test Recommendation
+
+  The test code consumes Kubernetes API as a serviceaccount/user with privileges:
+
+  - Create/Delete a namespace
+  - Create a ReplicaSet in the new namespace
+
+
+  Test setup
+
+  - Create a new namespace
+
+
+  Test steps  
+
+  - Create a ReplicaSet, in PodTemplate, without specifying `SecurityContext`, `ServiceAccountName`, without any volumes
+  - Poll the ReplicaSet and the namespace, and a new Pod should be created
+  - Create a ReplicaSet, in PodTemplate, specifying `privileged` as `true` in `SecurityContext`
+  - Poll the ReplicaSet, should got `ReplicaFailure` condition from `status`, and no Pod is created
+
+
+  Test teardown
+
+  - Delete the namespace
+
+#### Container Privilege Escalation
+
+  *Conformance Profile: Restricted*
+
+  Pod creation with `allowPrivilegeEscalation` set to `true` in `SecurityContext` (Pod or container level)
+  should fail if there's no explicit RBAC associated with the Pod service account
+  (`default` or explicitly specified in `ServiceAccountName`).
```

Comment:
I don't think we should go too far down the rabbit whole of testing pod security policy. If the goal is to test that there is a restricted PSP by default, I think we should just test that directly.

Reply:
I'm thinking a simple checking is more from implementation perspective, not reflecting the behavior or scenario. The test should make sure, in a behavioral way, all the pieces are setup properly, including PSP, RBAC etc.

Reply:
I mean we have e2e tests for PSP, RBAC, etc. If those tests don't provide sufficient coverage of the functionality, we should fix those tests. I think here you should just test 1) expected PSP is configured, and 2) PSP is enforced (a single check is sufficient).

Although actually, now that I say that I realize that you should be able to run a conformant cluster without using PSP - e.g. if you're enforcing that policy with a third party controller like OPA.